File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int goodNodes(TreeNode* root) {
        dfs(root, INT_MIN);
        return res;
    }
private:
    void dfs(TreeNode* node, int maxCurr) {
        if (maxCurr <= node->val) res++;
        
        if (node->left) {
            dfs(node->left, max(maxCurr, node->val));
        }
        if (node->right) {
            dfs(node->right, max(maxCurr, node->val));
        }
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        stack = [(root, float('-inf'))]
        res = 0
        while stack:
            node, max_curr = stack.pop()
            if max_curr <= node.val:
                res += 1
            if node.left:
                stack.append((node.left, max(max_curr, node.val)))
            if node.right:
                stack.append((node.right, max(max_curr, node.val)))
        return res